<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Detect marker panel from scRNA-seq profile â€¢ sc2marker</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Detect marker panel from scRNA-seq profile">
<meta property="og:description" content="More about what it does (maybe more than one line)
    Use four spaces when indenting paragraphs within the Description.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">sc2marker</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/vignette.html">vignette</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/CostaLab/sc2marker/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">

<div class="section level1">
<div class="page-header"><h1 id="sc2marker">sc2marker<a class="anchor" aria-label="anchor" href="#sc2marker"></a>
</h1></div>
<p>sc2marker is a toolkit for identification of cell specific markers for clusters of cells in a single cell RNA sequencing data set. In particular, sc2marker focus on markers that have valid antibodies to be used for either imaging or flow analyis.</p>
<div class="section level2">
<h2 id="install">Install<a class="anchor" aria-label="anchor" href="#install"></a>
</h2>
<p>You can install sc2marker with the following commands:</p>
<pre class="{r}"><code><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span>
<span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://devtools.r-lib.org/reference/remote-reexports.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"https://github.com/CostaLab/sc2marker"</span>, build_vignettes <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre>
</div>
<div class="section level2">
<h2 id="detect-markers-for-given-cell-cluster">Detect markers for given cell cluster<a class="anchor" aria-label="anchor" href="#detect-markers-for-given-cell-cluster"></a>
</h2>
<p>To run sc2marker you need to execute the following command, providing a clustered single cell data sets (as Seurat object), the cell type of interest and the antibody databased (IHC, ICC or Flow).</p>
<pre class="{r}"><code><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="va"><a href="https://github.com/CostaLab/sc2marker" class="external-link">sc2marker</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span>
<span class="va">nk.markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/Detect_single_marker.html">Detect_single_marker</a></span><span class="op">(</span><span class="va">mca.spleen</span>, id <span class="op">=</span> <span class="st">"NK cell(Spleen)"</span>, category <span class="op">=</span> <span class="st">"Flow"</span>, org <span class="op">=</span> <span class="st">"mouse"</span><span class="op">)</span></code></pre>
<p>You can display the results as a table with the command:</p>
<pre class="{r}"><code><span class="fu"><a href="reference/get_antibody.html">get_antibody</a></span><span class="op">(</span><span class="va">nk.markers</span>, org <span class="op">=</span> <span class="st">"mouse"</span><span class="op">)</span></code></pre>
<p>and you can generate ridge plot with the following command:</p>
<pre class="{r}"><code><span class="fu"><a href="reference/plot_ridge.html">plot_ridge</a></span><span class="op">(</span><span class="va">mca.spleen</span>, id <span class="op">=</span> <span class="st">"NK cell(Spleen)"</span>, genes <span class="op">=</span> <span class="va">nk.markers</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">9</span>,<span class="op">]</span><span class="op">$</span><span class="va">gene</span>, ncol <span class="op">=</span> <span class="fl">3</span>, assay <span class="op">=</span> <span class="st">"RNA"</span>, aggr.other <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></code></pre>
<p>To use customized gene set, you can run the following command. sc2marker will recognise genes and ignore the cases.</p>
<pre><code><span class="va">nk.markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/Detect_single_marker.html">Detect_single_marker</a></span><span class="op">(</span><span class="va">mca.spleen</span>, id <span class="op">=</span> <span class="st">"NK cell(Spleen)"</span>, category <span class="op">=</span> <span class="st">"Flow"</span>, geneset <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD19"</span>, <span class="st">"GeneA"</span>, <span class="st">"welcome2022"</span><span class="op">)</span>, org <span class="op">=</span> <span class="st">"mouse"</span><span class="op">)</span>
<span class="fu"><a href="reference/get_antibody.html">get_antibody</a></span><span class="op">(</span><span class="va">nk.markers</span>, rm.noab <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></code></pre>
</div>
<div class="section level2">
<h2 id="to-include-your-own-antibody-database">To include your own antibody database<a class="anchor" aria-label="anchor" href="#to-include-your-own-antibody-database"></a>
</h2>
<p>Example of self made antibody database can be found in <a href="Self_DB.csv">Self-made database</a>. Be sure to save the database within your R path and update the code above to reflect the file location in your system.</p>
<pre><code><span class="va">nk.markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/Detect_single_marker.html">Detect_single_marker</a></span><span class="op">(</span><span class="va">mca.spleen</span>, id <span class="op">=</span> <span class="st">"NK cell(Spleen)"</span>, category <span class="op">=</span> <span class="st">"Flow"</span>, org <span class="op">=</span> <span class="st">"mouse"</span>, self.db <span class="op">=</span> <span class="st">"/Path to Self_DB.csv/"</span><span class="op">)</span>
<span class="fu"><a href="reference/get_antibody.html">get_antibody</a></span><span class="op">(</span><span class="va">nk.markers</span>, self.db <span class="op">=</span> <span class="st">"/Path to Self_DB.csv/"</span>, org <span class="op">=</span> <span class="st">"mouse"</span><span class="op">)</span></code></pre>
</div>
<div class="section level2">
<h2 id="calculate-markers-for-all-cell-clusters">calculate markers for all cell clusters<a class="anchor" aria-label="anchor" href="#calculate-markers-for-all-cell-clusters"></a>
</h2>
<p>To calculate markers for all cell clusters, you can do by following command:</p>
<pre class="{r}"><code><span class="va">all.markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/Detect_single_marker_all.html">Detect_single_marker_all</a></span><span class="op">(</span><span class="va">mca.spleen</span>, category <span class="op">=</span> <span class="st">"Flow"</span>, org <span class="op">=</span> <span class="st">"mouse"</span><span class="op">)</span></code></pre>
<p>To Check T cell markers from results of all clusters, and get the antibody information, you can do following.</p>
<pre class="{r}"><code><span class="va">t.markers</span> <span class="op">&lt;-</span> <span class="va">all.markers</span><span class="op">[[</span><span class="st">"T cell(Spleen)"</span><span class="op">]</span><span class="op">]</span>
<span class="fu"><a href="reference/get_antibody.html">get_antibody</a></span><span class="op">(</span><span class="va">t.markers</span>, org <span class="op">=</span> <span class="st">"mouse"</span><span class="op">)</span></code></pre>
</div>
<div class="section level2">
<h2 id="generate-report">Generate report<a class="anchor" aria-label="anchor" href="#generate-report"></a>
</h2>
<p>To automatically generate the sc2marker report of all cell clusters, you can run following command:</p>
<pre class="{r}"><code><span class="fu"><a href="reference/generate_report.html">generate_report</a></span><span class="op">(</span><span class="va">mca.spleen</span>, <span class="va">all.markers</span>, fpath <span class="op">=</span> <span class="st">"."</span>, fname <span class="op">=</span> <span class="st">"mca.spleen"</span><span class="op">)</span></code></pre>
<p>Or you can only analysis subset of cell clusters (B cell and NK cell) and generate the report as following:</p>
<pre class="{r}"><code><span class="va">all.markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/Detect_single_marker_all.html">Detect_single_marker_all</a></span><span class="op">(</span><span class="va">mca.spleen</span>, category <span class="op">=</span> <span class="st">"Flow"</span>, clusters_to_detect <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Marginal zone B cell(Spleen)"</span>, <span class="st">"NK cell(Spleen)"</span><span class="op">)</span>, org <span class="op">=</span> <span class="st">"mouse"</span><span class="op">)</span>
<span class="fu"><a href="reference/generate_report.html">generate_report</a></span><span class="op">(</span><span class="va">mca.spleen</span>, <span class="va">all.markers</span>, fpath <span class="op">=</span> <span class="st">"."</span>, fname <span class="op">=</span> <span class="st">"mca.spleen"</span><span class="op">)</span></code></pre>
</div>
</div>
<div class="section level1">
<h1 id="reproduction">Reproduction<a class="anchor" aria-label="anchor" href="#reproduction"></a>
</h1>
<p>All code and R object needed to generate the results of sc2marker manuscript can be found in <a href="https://zenodo.org/record/5854754#.YeknMf7MKUm" class="external-link">zenodo</a>.</p>
<p>The scripts below reproduce individual analysis. Please note that some packages might require manual installation ( â€˜seurat-dataâ€™ and â€˜ggsciâ€™). Also, configure your R paths before running these scripts.</p>
<div class="section level3">
<h3 id="code-for-mca-spleen-mouse">Code for MCA Spleen (Mouse)<a class="anchor" aria-label="anchor" href="#code-for-mca-spleen-mouse"></a>
</h3>
<p>Example for MCA-Spleen in <a href="Example_MCA_Spleen.Rmd">Rmarkdown</a> format and in <a href="Example_MCA_Spleen.html">html</a> format.</p>
</div>
<div class="section level3">
<h3 id="code-for-human-bone-marrow-human-bm">Code for Human Bone marrow (human-BM)<a class="anchor" aria-label="anchor" href="#code-for-human-bone-marrow-human-bm"></a>
</h3>
<p><a href="https://github.com/satijalab/seurat-data" class="external-link">SeuratData</a> to load the R object. <a href="Example_human_BM.Rmd">Example_human_BM.Rmd</a></p>
</div>
<div class="section level3">
<h3 id="code-for-human-pbmclung-human-pbmclung">Code for Human PBMC&amp;lung (human-PBMC&amp;lung)<a class="anchor" aria-label="anchor" href="#code-for-human-pbmclung-human-pbmclung"></a>
</h3>
<p><a href="Example_human-PBMC&amp;lung.Rmd">Example_human_PBMC&amp;lung.Rmd</a></p>
</div>
<div class="section level3">
<h3 id="code-for-human-bone-marrow-human-pbmc">Code for Human Bone marrow (human-PBMC)<a class="anchor" aria-label="anchor" href="#code-for-human-bone-marrow-human-pbmc"></a>
</h3>
<p><a href="Example_human_PBMC.Rmd">Example_human_PBMC.Rmd</a></p>
<p>Also, you can find the original data of MCA, human-PBMC and human-PBMC&amp;lung and Stromal data as following: <a href="https://figshare.com/articles/dataset/MCA_DGE_Data/5435866" class="external-link">MCA</a>, <a href="https://archive.softwareheritage.org/browse/revision/1c7fcabb18a1971dc4d6e29bc3ed4f6f36b2361f/" class="external-link">human-PBMC&amp;lung</a>, <a href="https://atlas.fredhutch.org/nygc/multimodal-pbmc/" class="external-link">human-PBMC</a>, <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE128639" class="external-link">human-BM</a>, <a href="https://nicheview.shiny.embl.de/" class="external-link">niche stromal</a></p>
</div>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/CostaLab/sc2marker/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/CostaLab/sc2marker/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li>
<a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a>
</li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing sc2marker</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Ronghui Li <br><small class="roles"> Author, maintainer </small>  </li>
<li>Bella Banjanin <br><small class="roles"> Author </small>  </li>
<li>Rebekka Schneider <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0002-0749-1565" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li> Ivan Costa Gesteira <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0003-2890-8697" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>



  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Ronghui Li, Bella Banjanin, Rebekka Schneider, Ivan Costa Gesteira.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.3.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
